### ISSUE TITLE
400 error response on batch uploads

### ISSUE THREAD

**Issue by dthulin** [box/box-node-sdk#319]
- [ ] I have checked that the [SDK documentation][sdk-docs] and [API documentation][api-docs] doesn't solve my issue

### Description of the Issue
I'm running into a wall while trying to batch upload files with the node.js SDK. I can upload just fine when the code to upload is not wrapped in a batch request call, but as soon as I wrap it in a batch call I only get 400 responses from uploads. Documentation on this endpoint leaves something to be desired. Are uploads batchable with this SDK?

### Versions Used
Node SDK: // ## 1.12.1
Node.js: // v6.14.4 on a AWS Lambda function (editing and testing in cloud 9)

### Steps to Reproduce
 I have the following code: 
```
var box = sdk.getAppAuthClient('user','XXXXXXXXX'); // This number changed for github post
let parent = "55874501510"; // This number changed for github post
let fileToUpdate = "334131896150"; // This number changed for github post
let file1 = {"type":"Buffer","data":[84,104,105,115,32,105,115,32,97,32,98,117,102,102,101,114,32,101,120,97,109,112,108,101,46]}; 
let file2 = "Hello World!"; 
file1 = Buffer.from(file1);

//box.batch(); 
box.files.uploadFile(parent, 'Testy Test Test.txt', file1, (err, data) => { 
 if (err) { 
  console.log('uploadFile err',err); 
 } else {
  console.log(data.entries[0].id); 
 } 
}); 
box.files.uploadNewFileVersion(fileToUpdate, file2, (err, data) => { 
 if (err) { 
  console.log('uploadNewFileVersion err',err); 
 } else { 
  console.log(data.entries[0].id); 
 } 
});

/* 
box.batchExec(function(err, res) { 
 if(err) { 
  if (err instanceof TypeError) { 
   console.log('batch error hit:',err); 
   console.log('batch TypeError ,err.message',err.message); 
  } else { 
   console.log('batch error hit:',err); 
   console.log('batch err.response.body.code',err.response.body.code); 
  } 
 } else { 
  console.log('batch res', res); 
 } 
}); 
*/ 
```
When I run this in it's current state I console log the two file Ids as expected, however when I run it with all lines uncommented I get...

### Error Message, Including Stack Trace

```
2018-10-19 17:27:23.731 uploadFile err { Error: Unexpected API Response [400 Bad Request] (bad_request: "Bad Request")
    at Object.buildResponseError (/var/task/node_modules/box-node-sdk/lib/util/errors.js:57:23)
    at Object.buildUnexpectedResponseError (/var/task/node_modules/box-node-sdk/lib/util/errors.js:94:15)
    at /var/task/node_modules/box-node-sdk/lib/box-client.js:705:19
    at tryCatcher (/var/task/node_modules/bluebird/js/release/util.js:16:23)
    at Promise.successAdapter [as _fulfillmentHandler0] (/var/task/node_modules/bluebird/js/release/nodeify.js:23:30)
    at Promise._settlePromise (/var/task/node_modules/bluebird/js/release/promise.js:566:21)
    at Promise._settlePromise0 (/var/task/node_modules/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises (/var/task/node_modules/bluebird/js/release/promise.js:693:18)
    at Async._drainQueue (/var/task/node_modules/bluebird/js/release/async.js:133:16)
    at Async._drainQueues (/var/task/node_modules/bluebird/js/release/async.js:143:10)
    at Immediate.Async.drainQueues (/var/task/node_modules/bluebird/js/release/async.js:17:14)
    at runCallback (timers.js:672:20)
    at tryOnImmediate (timers.js:645:5)
    at processImmediate [as _immediateCallback] (timers.js:617:5)
  statusCode: 400,
  response: 
   { statusCode: 400,
     headers: {},
     body: 
      { type: 'error',
        status: 400,
        code: 'bad_request',
        help_url: 'http://developers.box.com/docs/#errors',
        message: 'Bad Request',
        request_id: '44tmnpfwe1oktg9j' } } }
2018-10-19 17:27:23.735 uploadNewFileVersion err { Error: Unexpected API Response [400 Bad Request] (bad_request: "Bad Request")
    at Object.buildResponseError (/var/task/node_modules/box-node-sdk/lib/util/errors.js:57:23)
    at Object.buildUnexpectedResponseError (/var/task/node_modules/box-node-sdk/lib/util/errors.js:94:15)
    at /var/task/node_modules/box-node-sdk/lib/box-client.js:705:19
    at tryCatcher (/var/task/node_modules/bluebird/js/release/util.js:16:23)
    at Promise.successAdapter [as _fulfillmentHandler0] (/var/task/node_modules/bluebird/js/release/nodeify.js:23:30)
    at Promise._settlePromise (/var/task/node_modules/bluebird/js/release/promise.js:566:21)
    at Promise._settlePromise0 (/var/task/node_modules/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises (/var/task/node_modules/bluebird/js/release/promise.js:693:18)
    at Async._drainQueue (/var/task/node_modules/bluebird/js/release/async.js:133:16)
    at Async._drainQueues (/var/task/node_modules/bluebird/js/release/async.js:143:10)
    at Immediate.Async.drainQueues (/var/task/node_modules/bluebird/js/release/async.js:17:14)
    at runCallback (timers.js:672:20)
    at tryOnImmediate (timers.js:645:5)
    at processImmediate [as _immediateCallback] (timers.js:617:5)
  statusCode: 400,
  response: 
   { statusCode: 400,
     headers: {},
     body: 
      { type: 'error',
        status: 400,
        code: 'bad_request',
        help_url: 'http://developers.box.com/docs/#errors',
        message: 'Bad Request',
        request_id: '44tmnpfwe1oktg9j' } } }
```
Any idea why I can't call the files.uploadFile and files.uploadNewFileVersion functions from within a batch without getting a 400 error response?

[sdk-docs]: ./docs
[api-docs]: https://developer.box.com/docs

**Comment by mattwiller** [comment]
@dthulin Uploads are not supported in the Batch API, which is why the API is giving you an error.

**Comment by dthulin** [comment]
@mattwiller Could you please add that to the batch documentation? Thanks!

**Comment by dthulin** [comment]
@mattwiller Also, what about downloads?

**Comment by mattwiller** [comment]
@dthulin Downloads are not supported via Batch API either; I've updated the documentation to reflect this.  Out of curiosity, what's your use case for wanting to batch those types of calls?

**Comment by dthulin** [comment]
@mattwiller Thanks so much, you might want to also add the 20 calls per batch limit too, IMHO. Also, though I think I can get and patch metadata in a batch, unless I'm mistaken adding metadata is not accessible via batch either. I'm sure there are more. I think I understand why they aren't batch-able, and I have no argument with why these calls aren't batch-able. (With the possible exception of adding metadata, but it might make more sense if I had a look under the hood of your systems.) It was more the lack of definitive documentation which, though somewhat understandable (it's a struggle we all face), was a bit frustrating.

As far as use cases goes, I must admit that I don't have a specific one right now, and I'm not certain if it will ever really be a thing. We like to keep things modular where we work, so we have an in-house custom API middle-ware that we built out on AWS which acts as gatekeeper for some of our 3rd party resources, allowing us to call from any system regardless of stack and get responses and error handling tailor made to our business processes. As an example we have a `findFolder(parent, folderName, createIfNotFound)` function at this endpoint. It uses your `box.folders.getItems(parent)`. It then searches entries for the `folderName`. If it can't be found and `createIfNotFound` is set to true, it then calls `box.folders.create(parent, folderName)`. There are trade-offs, but it's how we like it.

Since we need to use your product across our various systems, and your node sdk is currently more robust than your salesforce sdk, we decided that Box.com would be a good candidate to be housed in an AWS lambda function that resides at an one such API endpoint. I recently realized that we were hitting call limits due to not implementing the batch-able functionality. As a result I have been refactoring this utility endpoint from the ground up to allow us to use batchs. Now that I have done that, I have been going through and testing every endpoint that we have built out to ensure that it still functions as an independent call as well as part of a batch-able call. It was during this testing that I ran into problems. I have added comments in the code for this endpoint to reflect that these methods are not batch-able, and will have to add error handling to return an "action not allowed" response, but I wanted to make sure that it was indeed due to a lack of functionality on your part, as opposed to bad programming on my part.

**Comment by mattwiller** [comment]
@dthulin The metadata calls should all be batchable, but you may need to manually add the `Content-Type: application/json` header for those calls.  Some of those endpoints are very strict about that header, and headers are **not** inherited by each batch call, so from the API's point of view they don't have the header unless you specifically add it.  I'll make a note to ensure this works in the SDK.

One thing to keep in mind about the Batch API is that — except for very specific cases — each subrequest in a batch counts individually against rate limits.  So, if you try to batch up 20 `box.folders.getItems(...)` calls, that will still count as 20 requests for purposes of rate limiting.

The only cases where multiple calls are not counted against rate limits are:
* `box.folders.get(...)`
* `box.files.get(...)`
* `box.collaborations.get(...)`

**Comment by dthulin** [comment]
@mattwiller a few questions...
1. So, how do I add `Content-Type: application/json` to the header when I'm using your node SDK methods like this?
```
box.batch();
box.files.addMetadata(fileId, scope, template, metadata, (err, data) => {INSERT EPIC LOGIC HERE});
box.batchExec(function(err, res) {INSERT EPIC LOGIC HERE});
```
2. Are those 3 gets indeed the only calls that can be batched while only counting as 1 call?
3. Why is the batch documentation so limited? At first I thought the lack of documentation on batch calls was mildly irritating. Now I find it infuriating. I have spent soooo much time on this project since utilizing batches was recommended to me as a possible solution by your employees at BoxWorks. And while, yes I can find this info if I crawl through your forums: https://community.box.com/t5/Platform-and-Development-Forum/BOX-API-call-count/td-p/52854 I can also find conflicting info from your employees as well: https://github.com/box/box-windows-sdk-v2/issues/362#issuecomment-315875422 (note this is one of the few places on the whole internet where I could get the 20 calls max per batch limitation, but it's not in your documentation).
4. Can we get a list of the few calls that will reduce call Limits, the batch limit of 20, and other basic FAQs added to this section to save other users time and aggravation?

My knee jerk reaction was "why even have the calls batch-able then?!" but now that I've calmed down a bit, I wonder if it has saved me a small amount network traffic, and latency was a bit of a thing in one or two apps. I understand the need for rate limiting, but it would be great to have it documented. If the choke point is your back-end infrastructure, it would also be nice to have some sort of queue-able batch call that can run in the background when you have free resources, that don't count so heavily against the call Rate Limits.

My most intensive use case currently is a method that fires on a button click which initiates sharing with our client's end-users (~5 documents on average per folder for each end-user which averages ~200 folders per instance). We only fire this a couple of times a day, but it is a size-able amount of burst traffic. When our managed users finish preparing the documents, they hit the button which does the following: We check the internal directories with the working documents against a sharing directory. If the files are absent or have a newer modified date we bring them over. We make sure certain metadata is attached to both the original and the sharing version of the document. Finally, we ensure the proper external collaborators are invited to the sharing directory. Due to your Salesforce SDK being so far behind, and AWS lambdas only staying live up to 5 minutes currently, I had been hoping batch-ables would reduce the amount of time I'd have to spend Idling my JavaScript while waiting for my Rate Limits to clear.

**Comment by mattwiller** [comment]
@dthulin To answer your initial questions:

1. There isn't a way to set the header yourself when using the `addMetadata()` method — we'll have to patch the SDK.  In the meantime, you could always manually construct the call with the correct header using `client.post()`.
2. Yes, those are the only ones.  Batch API is __not__ intended to avoid rate limits; those rate limits are there to protect our backend infrastructure, and (generally) sending us 20 operations at one time vs. 20 operations in 20 different API calls does not change the amount of work the backend needs to do.  The primary use case for the Batch API as it exists today is to reduce network latency for users who would otherwise primarily be making API calls sequentially (e.g. if their programming language of choice does not offer non-blocking I/O as Node.js does).  In Node.js, API calls can easily be made in parallel, so Batch API is not as useful since it may not offer significant savings in the overall network latency.
3. The Batch API is __not__ currently considered Generally Available, and is largely undocumented for that reason.  As your comments indicate, there's still significant work to be done before this API is really useful for all the use cases people might have.  If you don't have a use case that takes advantage of the few cases where rate limits are avoided, I would not recommend using Batch API — just make the calls in parallel.
4. Since Batch API isn't publicly documented and is not considered stable, I'm hesitant to document it more in this SDK.
